ThreadSnapshot: redesign and cleanup

alexmenkov opened the pull request:
ThreadSnapshot: redesign and cleanup

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - Committer)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/loom.git pull/220/head:pull/220` \
`$ git checkout pull/220`

Update a local copy of the PR: \
`$ git checkout pull/220` \
`$ git pull https://git.openjdk.org/loom.git pull/220/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 220`

View PR using the GUI difftool: \
`$ git pr show -t 220`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/loom/pull/220.diff">https://git.openjdk.org/loom/pull/220.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/loom/pull/220#issuecomment-2885399475)
</details>

AlanBateman commented on src/java.base/share/classes/jdk/internal/vm/ThreadSnapshot.java:
```diff
@@ -153,32 +159,35 @@ Thread carrierThread() {
-        ELIMINATED_SCALAR_REPLACED,
-        ELIMINATED_MONITOR,
+        ELEMINATED_SCALAR_REPLACED,
+        ELEMINATED_MONITOR,
```
Did you mean to rename these?

AlanBateman commented on src/hotspot/share/classfile/javaClasses.cpp:
```diff
@@ -1933,43 +1949,33 @@ class GetThreadSnapshotClosure : public HandshakeClosure {
         oop park_blocker = java_lang_Thread::park_blocker(_java_thread());
         if (park_blocker != nullptr) {
-          _locks->push(LockInfo(depth, LockInfo::PARKING_TO_WAIT, park_blocker));
+          _blocker = LockInfo(LockInfo::PARKING_TO_WAIT, OopHandle(Universe::vm_global(), park_blocker));
```
At some point then maybe this could move out of detect_locks as it's nothing to do with the javaVFrame, instead parkBlocker is just a property of a thread and PARKING_TO_WAIT can be PARK_BLOCKER.

alexmenkov replied:
Added to TODO list

alexmenkov replied:
Fixed in update commit

alexmenkov commented on src/java.base/share/classes/jdk/internal/vm/ThreadSnapshot.java:
```diff
@@ -34,22 +34,32 @@ class ThreadSnapshot {
     private static final StackTraceElement[] EMPTY_STACK = new StackTraceElement[0];
     private static final ThreadLock[] EMPTY_LOCKS = new ThreadLock[0];
 
+    // filled by VM
```
Added to TODO list

alexmenkov replied:
Fixed in update commit

alexmenkov commented:
/integrate

alexmenkov commented:
Updated the fix:
- separated owned locks and blockers in Java and native;
- moved parkBlocker detection out of detect_locks();
- renamed PARKING_TO_WAIT to PARK_BLOCKER;
- implemented carrierThread functionality.

alexmenkov commented:
/integrate

AlanBateman commented:
/sponsor
